src/controller/datacontroller.js

```diff
@@ -306,6 +335,57 @@ export default class DataController {
 
 mix( DataController, EmitterMixin );
 
+// Traverses given DocumentFragment and searches elements which marks marker range. Founded element is removed from
+// DocumentFragment but path of this element is stored in Map.
+//
+// @param {module:engine/view/documentfragment~DocumentFragment} documentFragment Model DocumentFragment.
+// @returns {Object} Object with markers data and cleaned up document fragment.
+function extractMarkersFromModelFragment( documentFragment ) {
```

Comment:
To simplify this function and avoid refreshing walker, you could simply first find all elements to be extracted, and then remove them one by one after tree walker already did it's job:

```javascript
const markerStamps = new Set();
// ...
for ( const value of walker ) {
// ...
markerStamps.add( value.item )
// ...
}

for ( const stamp of markerStamps ) {
// ...
}
```

AFAIR, `Set` remembers the order in which items where stored.